README.md
# Nuxt GPT Content

[![npm version][npm-version-src]][npm-version-href]
[![npm downloads][npm-downloads-src]][npm-downloads-href]
[![License][license-src]][license-href]
[![Nuxt][nuxt-src]][nuxt-href]
![Status][status-src]

A Nuxt Content module capable of enhancing content creation via GPT APIs by querying a variety of models. Write content in your content files and include prompts that will be sent to Large Language Models (LLMs) such as OpenAI's GPT to generate and augment your content in order to quickly release new pages, posts, and any type of content you wish to manage.

- [✨ &nbsp;Release Notes](/CHANGELOG.md)
<!-- - [🏀 Online playground](https://stackblitz.com/github/your-org/nuxt-gpt-content?file=playground%2Fapp.vue) -->
<!-- - [📖 &nbsp;Documentation](https://example.com) -->

## Features

<!-- Highlight some of the features your module provide here -->
- 📃 Support for Markdown files and Front-Matter variables and prompts
- 🤖 Support for OpenAI models such as GPT3.5 and GPT4
- 🚀 Direct integration within Nuxt Content `^2.0.0` and Nuxt `^3.0.0` projects

## Quick Setup

1. Add `nuxt-gpt-content` dependency to your project

```bash
# Using pnpm
pnpm add -D nuxt-gpt-content

# Using yarn
yarn add --dev nuxt-gpt-content

# Using npm
npm install --save-dev nuxt-gpt-content
```

2. Add `nuxt-gpt-content` to the `modules` section of `nuxt.config.ts`

```js
export default defineNuxtConfig({
  modules: [
    'nuxt-gpt-content'
  ]
})
```

That's it! You can now use My Module in your Nuxt app ✨

## Configuration

1. Add a `gpt-content` configuration object to your `nuxt.config.ts` file

    ```js
    export default defineNuxtConfig({
    gptcontent: {
        active: true, //True by default but can be turned off
        contentModelProvider: 'openai', //`openai` or `none` for testing
        contentModelName: 'gpt-3.5-turbo', //Any supported OpenAI model name
        contentModelOptions: { //Any supported parameters to pass to the model config
            temperature: 0.8,
            n: 1,
            max_tokens: 250
        },
        commitHook: true, //Boolean or object containing commit details. Used to verify if a commit hook run after content files have been modified.
        saveContent: false //Should content files be updated and saved locally?
    },
    })
    ```
1. Ensure a you set an environment variable named `NUXT_OPENAI_API_KEY` with the value of you OpenAI API key.

## Development

```bash
# Install dependencies
npm install

# Generate type stubs
npm run dev:prepare

# Develop with the playground
npm run dev

# Build the playground
npm run dev:build

# Run ESLint
npm run lint

# Run Vitest
npm run test
npm run test:watch

# Release new version
npm run release
```

## License

[MIT License](https://github.com/GuillaumeCleme/nuxt-gpt-content/LICENSE)

<!-- Badges -->
[npm-version-src]: https://img.shields.io/npm/v/nuxt-gpt-content/latest.svg?style=flat&colorA=18181B&colorB=28CF8D
[npm-version-href]: https://npmjs.com/package/nuxt-gpt-content

[npm-downloads-src]: https://img.shields.io/npm/dm/nuxt-gpt-content.svg?style=flat&colorA=18181B&colorB=28CF8D
[npm-downloads-href]: https://npmjs.com/package/nuxt-gpt-content

[license-src]: https://img.shields.io/npm/l/nuxt-gpt-content.svg?style=flat&colorA=18181B&colorB=28CF8D
[license-href]: https://npmjs.com/package/nuxt-gpt-content

[nuxt-src]: https://img.shields.io/badge/Nuxt-18181B?logo=nuxt.js
[nuxt-href]: https://nuxt.com

[status-src]: https://img.shields.io/badge/Status-Experiemental-blue

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.4.0] - 2023-12-10
 
### Added
- Support for system prompts that help provide content to content models
### Changed
- N/A
### Fixed
- N/A

## [0.3.1] - 2023-12-10
 
### Added
- N/A
### Changed
- N/A
### Fixed
- Issue #3 - Commit hook options

## [0.3.0] - 2023-11-20
 
### Added
- Fix commit author details in commit hook
- Provide the ability to skip pushing to remote
- Implement module logger
- Update docs ith authentication details (was missing) 
### Changed
- N/A
### Fixed
- N/A

## [0.2.0] - 2023-11-20
 
### Added
- Minor package updates and unblock NPM publishing 
### Changed
- N/A
### Fixed
- N/A

## [0.1.0] - 2023-11-19
 
### Added
- Adding base functionality and support for OpenAI models.    
### Changed
- N/A
### Fixed
- N/A

test/fixtures/basic/nuxt.config.ts
import nuxtgptcontent from '../../../src/module'

export default defineNuxtConfig({
  modules: [
    nuxtgptcontent
  ],
  gptcontent: {
    contentModelProvider: 'openai',
    contentModelName: 'gpt-3.5-turbo',
    saveContent: true,
    active: true,
    commitHook: true
  },
})


src/runtime/commit.ts
import { execSync } from 'child_process'
import { ModuleOptions } from '../module';

import { useLogger } from '@nuxt/kit'

export function commitHook(options: ModuleOptions){

    const logger = useLogger('nuxt-gpt-content');

    try {
        executeCommand('git version');
    } catch (error) {
        throw new Error(`Could not find 'git', is it installed? Error: ${error}`);
    }

    const commitOptions = typeof options.commitHook == "object" ? options.commitHook : {};

    const gitStatus = executeCommand('git status --porcelain');

    if(gitStatus.toString().trim()){
        executeCommand('git add .')
        executeCommand(`git config user.name "${commitOptions.commitAuthorName ?? 'Nuxt GPT Content'}"`)
        executeCommand(`git config user.email "${commitOptions.commitAuthorEmail ?? 'no-reply@nuxt.com'}"`)
        executeCommand(`git commit -m "${commitOptions.commitMessage ?? 'Adding content changes'}"`)

        if(commitOptions.push){
            executeCommand('git push')
        }
    }
    else{
        logger.info("Nothing to commit");
    }
}

function executeCommand(command: string){
    const logger = useLogger('nuxt-gpt-content');
    logger.debug(`Executing: ${command}`);                
    try {
        return execSync(command);        
    } catch (error) {
        throw new Error(`Could not execute: '${command}'. Error: ${error}`);
    }
}

src/runtime/plugin.ts
import fs from 'fs';
import { parseMarkdown } from '@nuxtjs/mdc/dist/runtime'
import { ModuleOptions } from '../module';

import { defineNitroPlugin } from 'nitropack/dist/runtime/plugin'
import { useRuntimeConfig } from '#imports'

import OpenAIModel from './models/openai';
import LocalStub from './models/localstub';

import { useLogger } from '@nuxt/kit'

export default defineNitroPlugin((nitroApp) => {

    // @ts-ignore
    nitroApp.hooks.hook('content:file:beforeParse', async (file: {_id: string, body: string}, event) => {

        // @ts-ignore
        const config = useRuntimeConfig(event)?.gptcontent as ModuleOptions | undefined;
        const logger = useLogger('nuxt-gpt-content');

        //Return if not active
        if(!config?.active){
            logger.warn('Plugin set to inactive, skipping');
            return;
        }

        let parsedContent, matches;
        if(file._id.endsWith('.md')){
            //Front-matter keys
            parsedContent = await parseMarkdown(file.body);
            matches = file.body.match(/\${gpt_[_\w-]*}/) ?? [];
        }
        else{
            //Plain keys e.g. ${gpt_content=Write me a prompt}
            //matches = file.body.match(/\${gpt_[_\w-]*\=.*}/) ?? [];
            logger.warn('Only markdown files are supported in this version')
        }

        if(matches && parsedContent){
            for(const match of matches){
                const lookup = parsedContent.data[unwrapVariable(match)]

                let results;
                logger.info(`Invoking model: ${config?.contentModelProvider}`);
                switch (config?.contentModelProvider) {
                    case 'openai':
                        results = await new OpenAIModel(config).generateContent(lookup);
                        break;
                    default:
                        results = new LocalStub(config).generateContent(lookup);
                        break;
                }
    
                if(!results.results){
                    logger.error("Could not get content from model: " + results.error);
                }
                else{
                    file.body = file.body.replaceAll(match, results.results[0]);

                    if(config?.saveContent){
                        logger.info('Saving file content');
                        fs.writeFileSync(getFilePath(file), file.body);
                    }
                    else{
                        logger.info('Skipping file save');
                    }
                }
            }
        }
    })
})

function getFilePath(file: {_id: string, body: string}){
    return './' + file._id.replaceAll(':', '/');
}

function unwrapVariable(variable: string){
    return variable.substring(2, variable.length-1);
}

function splitVariable(variable: string) {
    variable = unwrapVariable(variable)
    return variable.slice(variable.indexOf('=')+1, variable.length);
}


src/runtime/models/openai.ts
import { ModuleOptions } from "../../module";
import { ModelResults, Model } from ".";
import OpenAI from 'openai';
import { resolve } from "path";
import fs from 'fs';
import { ChatCompletionMessageParam } from "openai/resources/chat/completions";


const openai = new OpenAI({
    apiKey: process.env.NUXT_OPENAI_API_KEY
});


export default class OpenAIModel implements Model {
    
    config: ModuleOptions;

    constructor(config: ModuleOptions){
        this.config = config;
    }

    async generateContent(message: string): Promise<ModelResults> {

        const messages: ChatCompletionMessageParam[] = [];
        if(this.config.contentModelSystemPrompt){
            const systemPromptPath = resolve(this.config.contentModelSystemPrompt);
            const systemPrompt = fs.readFileSync(systemPromptPath).toString();
            messages.push({ role: "system", content: systemPrompt });
        }

        messages.push({ role: "user", content: message });

        const chatCompletion = await openai.chat.completions.create({
            messages,
            model: this.config.contentModelName,
            ...this.config.contentModelOptions
        });
    
        if(!chatCompletion.choices[0].message.content){
            console.error("Could not get content from OpenAI: " + chatCompletion.choices[0].finish_reason);
            return {
                error: chatCompletion.choices[0].finish_reason
            }
        }
        else{
            return {
                results: [
                    chatCompletion.choices[0].message.content
                ]
            }
        }
    }   
}

src/runtime/models/localstub.ts
import { ModuleOptions } from "../../module";
import { Model, ModelResults } from ".";
import { resolve } from "path";
import fs from 'fs';
import { useLogger } from "@nuxt/kit";

export default class LocalStub implements Model {

    config: ModuleOptions;

    constructor(config: ModuleOptions){
        this.config = config;
    }

    generateContent(message: string): ModelResults {

        const logger = useLogger('nuxt-gpt-content');

        let systemPrompt;
        if(this.config.contentModelSystemPrompt){
            const systemPromptPath = resolve(this.config.contentModelSystemPrompt);
            systemPrompt = fs.readFileSync(systemPromptPath);
        }

        return {
            results: [
                `This content came form a local stub model. Received: [${message}] and system prompt: [${systemPrompt}]`
            ]
        }
    }
}

src/module.ts
import { defineNuxtModule, createResolver, useLogger } from '@nuxt/kit'
import { commitHook } from './runtime/commit'


// Module options TypeScript interface definition
export interface ModuleOptions {
    contentModelProvider: "openai" | "none"
    contentModelName: "gpt-3.5-turbo-1106" | "gpt-3.5-turbo" | "gpt-4" | "gpt-4-32k" | string
    contentModelOptions?: Record<string, any>,
    contentModelSystemPrompt?: string,
    commitHook: boolean | {
        commitAuthorEmail?: string,
        commitAuthorName?: string,
        commitMessage?: string,
        push?: boolean
    }
    saveContent: boolean,
    active: boolean
}

export default defineNuxtModule<ModuleOptions>({
    meta: {
        name: 'nuxt-gpt-content',
        configKey: 'gptcontent',
        compatibility: {
            // Semver version of supported nuxt versions
            nuxt: '^3.0.0'
        },
    },
    defaults: {
        active: true,
        contentModelProvider: 'openai',
        contentModelName: 'gpt-3.5-turbo',
        contentModelOptions: {
            temperature: 0.8,
            n: 1,
            max_tokens: 250,
        },
        commitHook: true,
        saveContent: false
    },
    setup(options, nuxt) {
        const logger = useLogger('nuxt-gpt-content')
        const resolver = createResolver(import.meta.url)

        //Inject Nitro Plugin
        nuxt.hook('nitro:config', async (nitroConfig) => {
            nitroConfig.plugins = nitroConfig.plugins ?? []
            nitroConfig.plugins?.push(resolver.resolve('./runtime/plugin'))

            nitroConfig.runtimeConfig = nitroConfig.runtimeConfig ?? {}
            nitroConfig.runtimeConfig.gptcontent = nitroConfig.runtimeConfig.gptcontent ?? options
        })

        //Register commit hook
        nuxt.hook('close', async () => {
            if (options.commitHook) {
                logger.info('Running commit hook');
                commitHook(options);
            }
            else{
                logger.info('Commit hook skipped');
            }
        })
    }
})